recipe_system/cal_service/file_getter.py

```diff
@@ -1,45 +1,112 @@
+import shutil
+import os.path
 import requests
 from requests.exceptions import HTTPError
 from requests.exceptions import Timeout
 from requests.exceptions import ConnectionError
 
-from gempy.utils import logutils
+from ..config import globalConf
+from gempy.utils.logutils import get_logger
+from .calrequestlib import generate_md5_digest
+
+# get_file_itterator is the only function in this module called externally
+# (from calrequestlib.py)
+
+def get_request(url, filename, calmd5):
+    """
+    This function was moved from calrequestlib.py as it belongs more naturally
+    in this module as it is the only thing that calls methods in this module.
+    Those methods where replaced with the CachedFileGetter class.
+    :param url: URL to fetch data from
+    :param filename: Filename to put data into
+    :param calmd5: md5 hexdigest of cal we are requesting
+    :return: None
+    """
+    cfg = CachedFileGetter()
+    cfg.get_request(url, filename, calmd5)
+
+class CachedFileGetter(object):
+    """
+    This class implements a file getter with an optional system- (or user-)
+    wide cache. To enable the cache, set system_calcache_dir to point at the
+    cache directory in the [calibs] section of your dragonsrc. If this is unset
+    or None, no caching is attempted. This cache is only really useful on
+    machines which run lots of instances of reduce in separate working
+    directories, notably for archive processing.
+
+    This class is not a singleton, and we don't store any cache concurrency data
+    in the class, we check file existence and compute md5 digests every time we
+    service a request.
+    """
+    def __init__(self):
+        self.cachedir = None
+        self.log = get_logger(__name__)
+
+        calconf = globalConf['calibs']
+        if calconf:
+            self.cachedir = calconf.get('system_calcache_dir')
+
+    def _fetchurltofile(self, url, filepath):
+        """
+        Fetch the contents of url and put them in a file at filepath
+        :param url: url to fetch
+        :param filepath: file to store contents in
+        :return: None
+        """
+        with open(filepath, 'wb') as fp:
+            r = requests.get(url, stream=True, timeout=10.0)
+            try:
+                r.raise_for_status()
+                for chunk in r.iter_content(chunk_size=None):
+                    fp.write(chunk)
+            except HTTPError as err:
+                raise GetterError(
+                    ["Could not retrieve {}".format(url), str(err)])
+            except ConnectionError as err:
+                raise GetterError(
+                    ["Unable to connect to url {}".format(url), str(err)])
+            except Timeout as terr:
+                raise GetterError(["Request timed out", str(terr)])
```

Comment:
```suggestion
            except HTTPError as err:
                raise GetterError(
                    ["Could not retrieve {}".format(url), str(err)]) from err
            except ConnectionError as err:
                raise GetterError(
                    ["Unable to connect to url {}".format(url), str(err)]) from err
            except Timeout as terr:
                raise GetterError(["Request timed out", str(terr)]) from terr
```

Reply:
Mainly suggesting this because it would capture the stack trace for the original error (in case it's relevant). I think leaving the message itself in the raised `GetterError` is fine, since it'll help inform in the original exception would be useful to look at.